### PR TITLE
Rework event generation

### PR DESCRIPTION
**Pull request by Aaron1011**
This PR makes a number of (interrelated) improvements to the event generation system:

* Previously, for methods overridden by other interfaces (think `getEntity`, `getLiving`, `getHuman`, etc), only one accessor and mutator would be generated. Due to the confusing way in which java determines which return type would be used, the method called wouldn't always be implemented. This PR generates accessors and mutators for all overridden methods - using only one field.

* Instead of having all generated event classes extend from `AbstractEvent`, the base class can be specified using the new `@ImplementedBy` annotation, or the event generator plugin system. This is used to implement methods which don't conform to the accessor/mutator pattern expected by the class generator (`getXXX`, `setXXX`, `isXXX`).

* The sponge event factory test has been extended to call all of the methods on event class, instead of just generating the classes. This ensures that all events generate as expected.

The event generator plugin system allows plugins or implementations to customize the resolution of event base classes, through registering an `EventFactoryPlugin`. This allows a different superclass to be chosen on a per-event basis, without introducing the requirement of a bytecode manipulation framework.

Several events have had methods modified or annotations added in order to correctly compile into bytecode.

### CODE REVIEW

**Comment by kenzierocks** [comment]
Pretty.

File: src/main/java/org/spongepowered/api/util/event/factory/AnnotationEventFactoryPlugin.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                }
+                for (Class<?> implInterfaces : scannedType.getInterfaces()) {
+                    queue.offer(implInterfaces);
+                }
+            }
+
+            if (implementedBy != null) {
+                return implementedBy.value();
+            }
+            throw new RuntimeException("Congratulations. You've attempted to create an event which doesn't have"
```
Comment (Mumfrey) [inline src/main/java/org/spongepowered/api/util/event/factory/AnnotationEventFactoryPlugin.java]: Need spaces where your strings concatenate because you're gonna end up with stuff like `doesn't havean @ImplementedBy` in the output.

**Comment by ryantheleach** [comment]
> The event generator plugin system allows plugins or implementations to customize the resolution of event base classes, through registering an EventFactoryPlugin. This allows a different superclass to be chosen on a per-event basis, without introducing the requirement of a bytecode manipulation framework.

This sounds cool, but what does it actually mean to a plugin developer?

**Comment by Mumfrey** [comment]
> This sounds cool, but what does it actually mean to a plugin developer?

It means you can raise custom events which cannot be fully generated by the class generator by specifying an abstract base class which implements the additional methods. The `EventFactoryPlugin` is capable of handling base-class resolution outside the scope of the normal annotation-based behaviour in the base system.

File: src/main/java/org/spongepowered/api/event/SpongeEventFactory.java [hunk 2]
```diff
@@ -171,17 +175,21 @@
 
     private static final FactoryProvider factoryProvider;
     private static final LoadingCache<Class<?>, EventFactory<?>> factories;
+    private static final Deque<EventFactoryPlugin> plugins;
 
     static {
         factoryProvider = new ClassGeneratorProvider("org.spongepowered.api.event.impl");
```
Comment (Mumfrey) [inline src/main/java/org/spongepowered/api/event/SpongeEventFactory.java]: Would it be preferable to put a marker interface in the target package and then use `Class.getPackage()` just to make this package declaration "soft"?

File: src/main/java/org/spongepowered/api/event/SpongeEventFactory.java [hunk 3]
```diff
@@ -171,17 +175,21 @@
 
     private static final FactoryProvider factoryProvider;
     private static final LoadingCache<Class<?>, EventFactory<?>> factories;
+    private static final Deque<EventFactoryPlugin> plugins;
```
Comment (Mumfrey) [inline src/main/java/org/spongepowered/api/event/SpongeEventFactory.java]: Much as I love `Deque`, does this really need to be one?

File: src/main/java/org/spongepowered/api/event/SpongeEventFactory.java [hunk 4]
```diff
@@ -189,6 +197,25 @@
     private SpongeEventFactory() {
     }
 
+    private static Class<?> getBaseClass(Class<?> event) {
+        Class<?> superClass = null;
+        for (EventFactoryPlugin plugin: plugins) {
+            superClass = plugin.resolveSuperClassFor(event, superClass);
+        }
+        return superClass;
```
Comment (Mumfrey) [inline src/main/java/org/spongepowered/api/event/SpongeEventFactory.java]: Maybe return a default `AbstractEvent` here if `superClass` is still null at this point?

File: src/main/java/org/spongepowered/api/util/event/factory/AnnotationEventFactoryPlugin.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * This is the base {@link EventFactoryPlugin}. If no other event factory plugins are able
+ * to handle the specified event, then the annotation on the event interface is used
+ * by this plugin to determine the superclass.
+ */
+public class AnnotationEventFactoryPlugin implements EventFactoryPlugin {
+
+    @Override
+    public Class<?> resolveSuperClassFor(Class<?> eventClass, Class<?> superClass) {
+        if (superClass == null) {
```
Comment (Mumfrey) [inline src/main/java/org/spongepowered/api/util/event/factory/AnnotationEventFactoryPlugin.java]: Why not short-circuit this check?

```java
if (superClass != null) {
    return;
}```

File: src/main/java/org/spongepowered/api/event/SpongeEventFactory.java (hunk 4 above)
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/event/SpongeEventFactory.java]: There should always be a superclass returned, either through a custom plugin, or by `AnnotationEventFactoryPlugin` reading the `@ImplementedBy` annotation from `Event`. If null *is* returned, it's most likely an error, and should be exposed as such.

File: src/main/java/org/spongepowered/api/event/SpongeEventFactory.java (hunk 2 above)
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/event/SpongeEventFactory.java]: How would the target package be discovered, since right now it's a different package from `SpongeEventFactory`.

This could probably go in a separate PR.
